openpgp-dsm/src/lib.rs

```diff
@@ -1129,6 +1165,320 @@ pub fn extract_cert(identifier: KeyIdentifier, cred: Credentials) -> Result<Cert
+            let pk: Key<PublicParts, SubordinateRole> = subkey
+                .sequoia_key.as_ref().context("unloaded subkey")?.clone().into();
+
+            let builder = SignatureBuilder::new(SignatureType::SubkeyBinding)
+                .set_key_validity_period(validity_period)?
+                .set_hash_algo(hash_algo)
+                .set_signature_creation_time(pk.creation_time())?
+                .set_key_flags(flags.clone())?;
+
+            let signature = pk.bind(&mut prim_signer, &cert, builder)?;
+            cert = cert.insert_packets(vec![Packet::from(pk), signature.into()])?;
+        }
+    }
+
+    let primary_key: Key<PublicParts, PrimaryRole> = primary
+                .sequoia_key.as_ref().context("unloaded subkey")?.clone().into();
+
+    if !new_signing_subkeys.is_empty() {
+        info!("key rotation: create embedded signature (signing-subkey signs primary)");
+        // To sign primary key
+        let mut subkey_signer = DsmAgent::new_signing_subkey_from_descriptor(
+            cred, &new_signing_subkeys[0].0.descriptor)?;
+        let embedded_signature = {
+            let subkey: Key<PublicParts, SubordinateRole> = new_signing_subkeys[0].0
+                .sequoia_key.as_ref().context("unloaded subkey")?.clone().into();
+
+            SignatureBuilder::new(SignatureType::PrimaryKeyBinding)
+                .set_key_validity_period(validity_period)?
+                .set_hash_algo(hash_algo)
+                .set_signature_creation_time(subkey.creation_time())?
+                .sign_primary_key_binding(&mut subkey_signer, &primary_key, &subkey)?
+        };
+
+        info!("key rotation: sign signing key");
+        {
+            let (subkey, flags) = (&new_signing_subkeys[0].0, &signing_subkey_flags);
+            let pk: Key<PublicParts, SubordinateRole> = subkey
+                .sequoia_key.as_ref().context("unloaded subkey")?.clone().into();
+
+            let builder = SignatureBuilder::new(SignatureType::SubkeyBinding)
+                .set_key_validity_period(validity_period)?
+                .set_hash_algo(hash_algo)
+                .set_signature_creation_time(pk.creation_time())?
+                .set_key_flags(flags.clone())?
+                .set_embedded_signature(embedded_signature)?;
+
+            let signature = pk.bind(&mut prim_signer, &cert, builder)?;
+            cert = cert.insert_packets(vec![Packet::from(pk), signature.into()])?;
+        }
+    }
+
+    info!("key rotation: bind new subkeys to primary key in DSM");
+    let links = KeyLinks {
+        parent: Some(primary.uid()?),
+        ..Default::default()
+    };
+    let link_update_req = SobjectRequest {
+        links: Some(links),
+        ..Default::default()
+    };
+
+    for (subkey, _) in &new_signing_subkeys {
+        dsm_client.__update_sobject(
+            &subkey.uid()?, &link_update_req, "bind new subkey to primary key"
```

Comment:
nit: "bind new signing subkey to primary key", and below "bind new encryption subkey to primary key"